Convert to JSR310

randomnicode opened the pull request:
A relatively simple refactor to change from legacy Date/Calendar classes to the modern JSR 310 java.time.* classes etc.

The number of files touched is pretty large, but most changes are actually simple drop-in replacements (please review)

PR is in two commits:
- First swaps the instances/logic of Date, Calendar, DateFormats with the java time pseudo equivalents
- Second deals with the JDBC mapping

The commit breakdown is deliberately such so as to allow us to revert the second commit once we get a JDBC 4.2 compliant driver and/or if we decide to use some aspects of an ORM.

Notes:
- Had to leave some instances of Date in there due to upstream method contracts
- Chose Instant over other Java Time classes as our primary replacement. Here are my reasons I'm choosing to write here for posterity:
  - We currently store data in the DB as `timestamp without timezone`, effectively epoch. We use the `Date` Java class which internally holds epoch elapsed.
  - Can't use `LocalDateTime`: Time will jump around if the server moves from place to place. You might have 8am for one timestamp and then transport the server (or migrate the db) to another timezone and get another 8am timestamp because `LocalDateTime.now()` will only give the local time. Additionally not backwards compatible with the current `Date` classes.
  - Can't use `OffsetDateTime`: We won't retain the timezone info and lose it without changing the columns to record timezones also (`timestamp with timezone`). I suspect it'll effectively become UTC anyway.
  - Thus we should use Instant and change it to whatever OffsetDateTime we want when we need it given the local timezone info. The data in the DB is always UTC

jvoisin commented on airsonic-main/src/main/java/org/airsonic/player/dao/AbstractDao.java:
```diff
@@ -69,11 +73,19 @@ protected String prefix(String columns, String prefix) {
     }
+    
+    protected static Object[] convertArgs(Object[] args) {
+        return args == null ? null : Stream.of(args).map(x -> (Object) ((x instanceof Instant) ? Timestamp.from((Instant) x) : x)).collect(Collectors.toList()).toArray();
```
Double ternaries are gross, wouldn't it be a bit clearer to split this on different lines?

randomnicode replied:
I could use Optionals too, I just think this is cleaner, otherwise we'll have to use block statements with returns inside the `map()`

jvoisin commented on airsonic-main/src/main/java/org/airsonic/player/service/MediaScannerService.java:
```diff
@@ -49,7 +54,9 @@
     private boolean scanning;
-    private Timer timer;
+    
+    private ScheduledExecutorService scheduler;
```
It seems that you mixed this PR with #1376 :)

randomnicode replied:
No, this isn't a mixup.

This is how GitHub deals with commits on successive PRs built on each other (please compare commit hashes for this PR and https://github.com/airsonic/airsonic/pull/1376 ). This PR is a successor of https://github.com/airsonic/airsonic/pull/1376 and built on top of it.

Once you merge https://github.com/airsonic/airsonic/pull/1376, the effective change in this PR is the remaining commit hashes that it doesn't share with https://github.com/airsonic/airsonic/pull/1376 (i.e. everything excluding https://github.com/airsonic/airsonic/pull/1377/commits/87285668606fa3ee6be23aecd7be012e59aeea8d which is part of https://github.com/airsonic/airsonic/pull/1376)

This is also the reason you shouldn't change commit hashes on PRs when merging them on master such as what happened in https://github.com/airsonic/airsonic/pull/1357 and actually even a recently merged one here https://github.com/airsonic/airsonic/pull/1353. It'll result in immediate conflicts in subsequent PRs built on top of another PRs because Git will find no commit history for the hashes in one PR, and immediately declare a merge conflict for all subsequent PRs built on the merged PR that has had its commit history changed. Additionally , the local branches won't recognize they've been merged and refuse to delete unless you force delete them (-D)

jvoisin commented on airsonic-main/src/main/java/org/airsonic/player/dao/UserDao.java:
```diff
@@ -164,7 +165,7 @@ public void deleteUser(String username) {
     public void updateUser(User user) {
         String sql = "update " + getUserTable() + " set password=?, email=?, ldap_authenticated=?, bytes_streamed=?, bytes_downloaded=?, bytes_uploaded=? " +
                 "where username=?";
-        getJdbcTemplate().update(sql, encrypt(user.getPassword()), user.getEmail(), user.isLdapAuthenticated(),
```
Why is this not needed anymore?

randomnicode replied:
Replaced by an `update()` call to the AbstractDAO parent, which does the same thing, but a few additional things:
- logging
- timing
- argument sanitization/conversion

Literally every other Dao class uses the AbstractDAO, this class was the only exception which calls the JdbcTemplate directly. It probably flew under the radar.

It now needs to be converted to be like every other Dao class because we have important timestamp conversions happening in AbstractDAO

muff1nman commented on airsonic-main/src/main/java/org/airsonic/player/service/playlist/XspfPlaylistExportHandler.java:
```diff
@@ -39,7 +40,7 @@ public SpecificPlaylist handle(int id, SpecificPlaylistProvider provider) {
         newPlaylist.setTitle(playlist.getName());
         newPlaylist.setCreator("Airsonic user " + playlist.getUsername());
-        newPlaylist.setDate(new Date());
+        newPlaylist.setDate(Date.from(Instant.now())); //TODO switch to Instant upstream
```
Yep good call.

randomnicode commented:
Hmmm, I don't know if you confused PRs or meant to post this for a different PR, but I don't think the issue you're having is related to this PR. Your workspace is likely polluted with something else you were working on. Try rerunning on a different VM or a fresh install.

- I'm unable to duplicate it. My application comes up just fine.
- It would've shown up in the test logs at the very least which it doesn't (please see the test logs)
- The error you posted is claiming it's unable to find a particular table, but this PR doesn't touch any table DDLs. The error is wholly unrelated.

I do however see some new addition of an ObjectMapper on master with serializations going on for a time class for which I'll need to register  a Jackson module.

randomnicode commented:
@muff1nman Hmmm, I don't know if you confused PRs or meant to post this for a different PR, but I don't think the issue you're having is related to this PR. Your workspace is likely polluted with something else you were working on. Try rerunning on a different VM or a fresh install.

- I'm unable to duplicate it. My application comes up just fine.
- It would've shown up in the test logs at the very least which it doesn't (please see the test logs)
- The error you posted is claiming it's unable to find a particular table, but this PR doesn't touch any table DDLs. The error is wholly unrelated.

I do however see some new addition of an ObjectMapper on master with serializations going on for a time field for which I'll need to register  a Jackson module.

muff1nman commented:
Okay I'll check again.

randomnicode commented on airsonic-main/src/main/java/org/airsonic/player/service/MediaFileService.java:
```diff
@@ -373,14 +374,14 @@ public void populateStarredDate(List<MediaFile> mediaFiles, String username) {
 
         // Check timestamps.
-        if (parent.getChildrenLastUpdated().getTime() >= parent.getChanged().getTime()) {
+        if (parent.getChildrenLastUpdated().isAfter(parent.getChanged())) {
```
This is a bug. we're not checking the '=' condition, only the '>' condition. Which actually matters a lot.

fxthomas commented on airsonic-main/src/main/java/org/airsonic/player/dao/AbstractDao.java:
```diff
@@ -69,11 +73,19 @@ protected String prefix(String columns, String prefix) {
+    }
+    
+    protected static Map<String, Object> convertArgs(Map<String, Object> args) {
+        return args == null ? null : args.entrySet().parallelStream().map(x -> (x.getValue() instanceof Instant) ? Pair.of(x.getKey(), Timestamp.from((Instant) x.getValue())) : x).collect(Collectors.toMap(x -> x.getKey(), x -> x.getValue()));
```
I do think it would be cleaner to have a separate method for converting single db objects... Then you would have to use a big lambda inside the `map`, it's hard to read.

randomnicode replied:
The way I view it, I think most of this is temporary plumbing code until we get (a) a more recent jdbc compliant driver (than hsqldb 1.8) that supports these more recent additions out of the box, and (b) move to a more pliable sql management library that'll do most of this work for us instead of relying on jdbc templates.

This stuff is here so people who want to remain on older versions can also enjoy the benefits, thus we need to provide these adapters to bridge them between old and new. As such, short and succinct works well enough, without putting more effort in.

I'll separate it out into multiple lines for readability, but to my lambdafied eyes, it's a simple stream function that checks every map entry to see if the value is an Instant, and if it is, it converts it to a Timestamp, and reassociates it with the same key. Otherwise it keeps the entry as is.

fxthomas replied:
Ah, I did not know about this being needed because of HSQLDB 1.8. Then let's add a `FIXME` comment so that we remember why we need this right now!

fxthomas commented on airsonic-main/src/main/java/org/airsonic/player/dao/AbstractDao.java:
```diff
@@ -69,11 +73,19 @@ protected String prefix(String columns, String prefix) {
+        return args == null ? null : Stream.of(args).map(x -> (Object) ((x instanceof Instant) ? Timestamp.from((Instant) x) : x)).collect(Collectors.toList()).toArray();
+    }
+    
+    protected static Map<String, Object> convertArgs(Map<String, Object> args) {
```
I'd name the method better than `convertArgs` for example `convertToDatabaseTypes`?

randomnicode replied:
👍 I'll rename

fxthomas commented on airsonic-main/src/main/java/org/airsonic/player/util/Util.java:
```diff
@@ -117,10 +119,11 @@ public static void setContentLength(HttpServletResponse response, long length) {
+    private static ObjectMapper objectMapper = new ObjectMapper()
+            .findAndRegisterModules()
+            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+            .configure(DeserializationFeature.READ_DATE_TIMESTAMPS_AS_NANOSECONDS, false)
```
If I understand correctly, this changes all DB operations to not care about nanoseconds anymore, right? If so, I'm okay with this, we don't gain a lot by storing nanoseconds.

randomnicode replied:
Not quite. This objectMapper is actually not used to serialize to the DB.

We seem to be using it for converting MediaLibraryStatistics to a Map<String, String> for storing as commit metadata in the lucene index whenever the scan data is committed. I believe that is a recent addition in #1262 

This is needed primarily for the reason that without this, deserialization reads whole numbers as number of seconds instead of number of milliseconds. Decimals will still be read at nanosecond resolution because that's not ambiguous

randomnicode commented:
> `javax.el.ELException: Cannot convert [2019-11-02T17:26:27.454Z] of type [class java.time.Instant] to [class java.util.Date]
	at org.apache.el.lang.ELSupport.coerceToType(ELSupport.java:576)`

Seems like a jstl related jsp error. Thanks for pointing it out, I'll take a look at what we can do about it.

randomnicode commented:
@fxthomas This should fix the jstl bugs mentioned. Any other issues you notice?

If not, we should think about merging this one. The PR backlog is starting to become pretty large here. And I have some more code I need to push out on top of this one (aside from the Media scanner parallelization)

fxthomas commented:
Sure, we'll merge this as soon as it's ready.

I did not see any crash, everything seems to work fine on my instance; now that it's usable I'll leave that running for a day or two and we can merge it.

Minor nitpick: going to a Playlist page displays fine but writes this in the logs:
```
ERROR --- o.d.impl.DefaultConverterManager         : No converter found for 'java.time.Instant'
```

fxthomas commented:
Ah, one more ; when going to More > Shuffle Radio:

```
java.time.temporal.UnsupportedTemporalTypeException: Unsupported unit: Months
	at java.base/java.time.Instant.plus(Instant.java:861)
	at java.base/java.time.Instant.minus(Instant.java:978)
	at org.airsonic.player.controller.RandomPlayQueueController.handleRandomPlayQueue(RandomPlayQueueController.java:133)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:133)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:97)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:854)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:765)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:967)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:872)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.airsonic.player.security.JWTRequestParameterProcessingFilter.doFilter(JWTRequestParameterProcessingFilter.java:62)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.airsonic.player.filter.MetricsFilter.doFilter(MetricsFilter.java:30)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.airsonic.player.filter.RequestEncodingFilter.doFilter(RequestEncodingFilter.java:45)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.airsonic.player.filter.ParameterDecodingFilter.doFilter(ParameterDecodingFilter.java:57)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.airsonic.player.filter.BootstrapVerificationFilter.doFilter(BootstrapVerificationFilter.java:55)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter(RememberMeAuthenticationFilter.java:158)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.airsonic.player.security.RESTRequestParameterProcessingFilter.doFilter(RESTRequestParameterProcessingFilter.java:90)
```

randomnicode commented:
@fxthomas Nice! I've fixed both:
- turns out Instant doesn't support units greater than a day being added though it has the api for it. went around it.
- fixed the dwr issue: just added a custom converter. it's too old a lib to modify it.
- fixed a potential bug in formatting Instants. mostly affecting podcasts and playlist creation. I've updated our podcast date parsing to just use the spec instead of our own custom formatter

fxthomas commented:
Great! Let me merge/run things locally one more time, and I think this should be ready to go.

fxthomas commented:
Ah, well More > Shuffle Radio again, sorry... :cry:

Potentially this kind of thing could impact other places that pass nulls to parametrized queries?

```
java.lang.NullPointerException
	at java.base/java.util.Objects.requireNonNull(Objects.java:222)
	at java.base/java.util.stream.Collectors.lambda$uniqKeysMapAccumulator$1(Collectors.java:178)
	at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.HashMap$EntrySpliterator.forEachRemaining(HashMap.java:1748)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:952)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:926)
	at java.base/java.util.stream.AbstractTask.compute(AbstractTask.java:327)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinTask.doInvoke(ForkJoinTask.java:408)
	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:736)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateParallel(ReduceOps.java:919)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at org.airsonic.player.dao.AbstractDao.convertToDBTypes(AbstractDao.java:88)
	at org.airsonic.player.dao.AbstractDao.namedQueryWithLimit(AbstractDao.java:128)
	at org.airsonic.player.dao.MediaFileDao.getRandomSongs(MediaFileDao.java:595)
	at org.airsonic.player.dao.MediaFileDao$$FastClassBySpringCGLIB$$15ad9359.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
```

randomnicode commented:
Excellent, keep'em coming. I'll look to see how we can resolve this one

randomnicode commented:
@fxthomas  Excellent, keep'em coming. Resolved this one, and also checked that streams can collect into lists with nulls just fine (it's just maps that have problems).

I am curious though about why we're passing null args to a db named parameter. Maybe checking = null?

randomnicode commented:
Going to close this and dependent PRs and move it to another repo. Thanks for the comments and the code audit @fxthomas 

Good luck
